leshan-client-demo/src/main/java/org/eclipse/leshan/client/demo/LeshanClientDemo.java

```diff
@@ -745,6 +748,9 @@ public void handshakeFailed(Handshaker handshaker, Throwable error) {
         builder.setBootstrapAdditionalAttributes(bsAdditionalAttributes);
         final LeshanClient client = builder.build();
 
+        client.addObserver(
+            new ClientShutdownOnUnexpectedErrorObserver(client, myDevice, randomTemperatureSensor));
```

Comment:
This should be done on `LeshanClient` not `LeshanClientDemo`  or I missed something ?
(https://github.com/eclipse/leshan/issues/933#issuecomment-730848056)

Reply:
Yes, that's a point. I think it would be better to implement this shutdown observer mechanism on `LeshanClient`, but it breaks the backward compatibility of the client, so I delegated this behavior to the client of the client.

What do you think about this?

Reply:
For `master`, we don't care about backward compatibility question and should just focus on what could be the better behavior. (of course this doesn't mean we can break everything without any reason :sweat_smile: )

For `1.x`, we should try to not break the API and behavior but here we could almost consider this as a bug ... so changing this behavior doesn't disturb me. 

WDYT ?

Reply:
Yeah, I'm on the same page.
I also think it would be better to implement this on the LeshanClient.
I was scared to break the backward compatibility way too much, but it seems no problem. I'll do this.

One more thing. Would it useful to support a method to prevent shutting down even if an unexpected error has occurred on the LeshanClient? For example `LeshanClient#doNotDestroyOnUnexpectedError()`.

Reply:
I would not add it for now. Let's see if someone need it later.

Reply:
I still don't see what could be the use case. As task in `DefaultRegistrationEngine` are stopped. :thinking:

Reply:
yeah, my suggestion was for just in case. It looks there is no reason to support that. Thank you.

Reply:
fixed at 7e6d029